fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/query/plan/synthetic/SyntheticRecordPlannerTest.java

```diff
@@ -1036,6 +1042,363 @@ public void aggregateJoinIndex() throws Exception {
+                    .setCity("Toronto")
+                    .build();
+            recordStore.saveRecord(customer);
+
+            TestRecordsJoinIndexProto.OrderWithHeader order = TestRecordsJoinIndexProto.OrderWithHeader.newBuilder()
+                    .setHeader(TestRecordsJoinIndexProto.Header.newBuilder().setZKey(1).setRecId("23"))
+                    .setOrderNo(10)
+                    .setQuantity(23)
+                    .setCustRef(TestRecordsJoinIndexProto.Ref.newBuilder().setStringValue("i:1"))
+                    .build();
+            recordStore.saveRecord(order);
+
+            try (RecordCursor<IndexEntry> cursor = recordStore.scanIndex(joinIndex, IndexScanType.BY_VALUE, TupleRange.ALL, null, ScanProperties.FORWARD_SCAN)) {
+                List<IndexEntry> entries = cursor.asList().get();
+                assertEquals(1, entries.size());
+                final IndexEntry indexEntry = entries.get(0);
+                assertEquals("Scott", indexEntry.getKey().getString(0), "Incorrect customer name");
+                assertEquals(10L, indexEntry.getKey().getLong(1), "Incorrect order number");
+            }
+
+            // Update the customer name
+            TestRecordsJoinIndexProto.CustomerWithHeader customerWithNewName = customer.toBuilder()
+                    .setName("Alec")
+                    .build();
+            recordStore.saveRecord(customerWithNewName);
+
+            try (RecordCursor<IndexEntry> cursor = recordStore.scanIndex(joinIndex, IndexScanType.BY_VALUE, TupleRange.ALL, null, ScanProperties.FORWARD_SCAN)) {
+                List<IndexEntry> entries = cursor.asList().get();
+                assertEquals(1, entries.size());
+                final IndexEntry indexEntry = entries.get(0);
+                assertEquals("Alec", indexEntry.getKey().getString(0), "Incorrect customer name");
+                assertEquals(10L, indexEntry.getKey().getLong(1), "Incorrect order number");
+            }
+
+            // Update the order number
+            TestRecordsJoinIndexProto.OrderWithHeader orderWithNewNumber = order.toBuilder()
+                    .setOrderNo(42)
+                    .build();
+            recordStore.saveRecord(orderWithNewNumber);
+
+            try (RecordCursor<IndexEntry> cursor = recordStore.scanIndex(joinIndex, IndexScanType.BY_VALUE, TupleRange.ALL, null, ScanProperties.FORWARD_SCAN)) {
+                List<IndexEntry> entries = cursor.asList().get();
+                assertEquals(1, entries.size());
+                final IndexEntry indexEntry = entries.get(0);
+                assertEquals("Alec", indexEntry.getKey().getString(0), "Incorrect customer name");
+                assertEquals(42L, indexEntry.getKey().getLong(1), "Incorrect order number");
+            }
+
+            // Insert an order with no associated customer
+            TestRecordsJoinIndexProto.OrderWithHeader orderWithNoCustomer = TestRecordsJoinIndexProto.OrderWithHeader.newBuilder()
```

Comment:
Not a priority really, but these feel like we would want a separate test for these cases, so that we can see which parts of this test fail. Otherwise, the test itself could fail for a lot of different reasons that are hard to parse.

Reply:
Sure. I think my preference would be to do this in a separate PR. I've filed this issue to track it: #1887